src/api.coffee

```diff
@@ -47,15 +48,53 @@ module.exports = (application) ->
 		application.update(req.body.force)
 		res.sendStatus(204)
 
-	unparsedRouter.post '/v1/reboot', (req, res) ->
+	parsedRouter.post '/v1/reboot', (req, res) ->
+		force = req.body.force
 		utils.mixpanelTrack('Reboot')
-		request.post(config.gosuperAddress + '/v1/reboot')
-		.pipe(res)
+		knex('app').select()
+		.map (theApp) ->
+			Promise.using application.lockUpdates(theApp.appId, force), ->
+				# There's a slight chance the app changed after the previous select
+				# So we fetch it again now the lock is acquired
+				knex('app').select().where({ appId: theApp.appId })
+				.then ([ app ]) ->
+					application.kill(app, removeContainer: false) if app?
+		.then ->
+			new Promise (resolve, reject) ->
+				request.post(config.gosuperAddress + '/v1/reboot')
+				.on 'error', reject
+				.on 'response', -> resolve()
+				.pipe(res)
+		.catch (err) ->
+			if err instanceof application.UpdatesLockedError
+				status = 503
+			else
+				status = 500
+			res.status(status).send(err?.message or err or 'Unknown error')
 
-	unparsedRouter.post '/v1/shutdown', (req, res) ->
+	parsedRouter.post '/v1/shutdown', (req, res) ->
+		force = req.body.force
 		utils.mixpanelTrack('Shutdown')
-		request.post(config.gosuperAddress + '/v1/shutdown')
-		.pipe(res)
+		knex('app').select()
+		.map (theApp) ->
+			Promise.using application.lockUpdates(theApp.appId, force), ->
+				# There's a slight chance the app changed after the previous select
+				# So we fetch it again now the lock is acquired
+				knex('app').select().where({ appId: theApp.appId })
+				.then ([ app ]) ->
+					application.kill(app, removeContainer: false) if app?
+		.then ->
+			new Promise (resolve, reject) ->
+				request.post(config.gosuperAddress + '/v1/shutdown')
+				.on 'error', reject
```

Comment:
I'd prefer `.on('error', reject)`, since when I first read this I thought it was `.on 'error', -> reject` due to the lack of parens